### DISCUSSION TITLE
Could there be an option to change how beatmaps are sorted in song select?

### DISCUSSION THREAD

**Discussion by Squaduck** [ppy/osu#27260] [Ideas] [answered]
I've gotten used to the way beatmap sorting worked in previous releases and built up an idea of where beatmaps are in relation to one another.
With the new sorting ([Symbols][A-Z][a-z]) as compared to the old sorting ([Symbols][case-insensitive A-Z]) that's all gone.

Could there be an option to change how string sorting works in beatmap selection? 
(That's a genuine question, I have no clue how lazer works.)

From looking at the PR that changed it (#26895), it would just be setting an enum to a different value to change it.

Or better yet, have a string comparer delegate in the `FilterCriteria` class to call instead of `string.Compare` to accommodate the solution proposed in https://github.com/ppy/osu/pull/26895#issuecomment-1947891093 as well.

(Be warned that I have no clue what I'm doing and have no idea about preferred code styles)
It would be something vaguely like...

A definition for a StringComparer delegate:
```cs
public delegate int StringComparer(string a, string b);
```
A class of string comparers to switch between: (This is the thing I'm least sure about. I have no clue where you'd put this.)
```cs
public static class StringComparers
{
	// Method proposed by frenzibyte in https://github.com/ppy/osu/pull/26895#issuecomment-1947891093
	// Except I just switched a and b instead of inverting it
	public static int OrdinalIgnoreCase_Plus_Ordinal(string a, string b)
	{
		int result = string.Compare(a, b, StringComparison.OrdinalIgnoreCase);
		if (result == 0)
			result = string.Compare(b, a, StringComparison.Ordinal);
		return result;
	}

	public static int Ordinal(string a, string b) => string.Compare(a, b, StringComparison.Ordinal);

	public static int InvariantCulture(string a, string b) => string.Compare(a, b, StringComparison.InvariantCulture);
}
```
Then a `StringComparer` field in `FilterCriteria`: (Also kinda sketchy. don't really know if it belongs here.)
```diff
public class FilterCriteria
{
...
+++ public StringComparer StringComparer = StringComparers.Ordinal;
...
```
Then finally using that instead of just `string.Compare`:
```diff
public class CarouselBeatmapSet : CarouselGroupEagerSelect
{
...
    public override int CompareTo(FilterCriteria criteria, CarouselItem other)
    {
    ...
        switch (criteria.Sort)
        {
            default:
            case SortMode.Artist: // and all other strings
+++             comparison = criteria.StringComparer(BeatmapSet.Metadata.Artist, otherSet.BeatmapSet.Metadata.Artist);
---             comparison = string.Compare(BeatmapSet.Metadata.Artist, otherSet.BeatmapSet.Metadata.Artist, StringComparison.Ordinal);
...
```

I have no clue how options work in lazer, so I don't know how hard that would be to implement. 
I'll leave that to the people who know what they are doing. I know just enough C# to be dangerous.

This would also probably work to implement reverse sorting as an option.
But I also think that this would make it a bit over-complicated. IDK. 
You could also just add an option for the enum or just do nothing at all.

**TLDR**: I don't like the new sorting, I think it would be cool to have it as an option, just changing the enum should be plenty, but I kinda like frenzibyte's idea and I have a probably over-complicated idea that could make that work. 

I'm open to opinions and advice.

**Comment by peppy** [comment, marked as answer]
WIll be fixed in the next release.
